src/pipert2/core/base/message.py

```diff
@@ -1,2 +1,117 @@
+from src.pipert2.core.base.payload import Payload
+
+import collections
+import time
+import pickle
+
+
 class Message:
-    pass
+    """The Message is a wrapper for information that passes through the pipe.
+    The information cannot pass alone and it passes inside the Message object during its entire stay.
+    The Message is not exposed to the user and is used as a tool in the system that helps to route the information
+    and remember its transit history and later more.
+
+    """
+
+    counter = 0
+
+    def __init__(self, data: collections.Mapping, source_address: str):
+        """
+        Args:
+            data: Data that the message will hold.
+            source_address: Where the message was created. # TODO - Maybe remove this arg
+
+        Attributes:
+            payload (Payload): The payload that managing the data.
+            source_address (str): Where the Message was created
+            history (collections.OrderedDict): Transition history of the Message between the routines
+            id (str): Unique id for the Message object.
+
+        """
+
+        self.payload: Payload = Payload(data)
+
+        self.source_address = source_address
+        self.history = collections.OrderedDict()
+        self.id = f"{self.source_address}_{Message.counter}"
+
+        Message.counter += 1
+
+    def update_data(self, data: collections.Mapping):
+        """Update the data the message containing
+
+        Args:
+            data (collections.Mapping): dictionary containing the data
+
+        """
+
+        self.payload.data = data
+
+    def get_data(self) -> collections.Mapping:
+        """Get the data in the message
+
+        Returns:
+            Dictionary with the message data
+
+        """
+
+        if self.payload.encoded:
+            self.payload.decode()
+        return self.payload.data
+
+    def record_entry(self, routine_name) -> None:
+        """Records the timestamp of the message's entry into a component.
+
+        Args:
+            routine_name: the name of the routine that the message entered.
+
+        """
+
+        self.history[routine_name] = time.time()
```

Comment:
Maybe do it the other way around as in: `self.history[time.time()] = routine_name`?
Sense `self.history` is an `OrderedDict` won't it arrange itself according to ASCII values and it will mess up the order the message was going through the pipe?

Reply:
Ordered dict remember the order of the entries so i decided to save it like that and it won't mess up the order.